### PR TITLE
port: Register missingProperties custom function to get all variables the template contains

### LINKED ISSUE #3849: port: Register `missingProperties` custom function to get all variables the template contains (#5518)
The changes in [Register `missingProperties` custom function to get all variables the template contains (#5518)](https://github.com/microsoft/botbuilder-dotnet/pull/5518) may need to be ported to maintain parity with `microsoft/botbuilder-dotnet`.

<blockquote>
Fixes #3531

## Description
Register a new custom function: `missingProperties`, which could get all the variables the template contains.
</blockquote>

Please review and, if necessary, port the changes.

### PR DESCRIPTION
**Pull request by Danieladu**
Fixes #3849

## Description
Register a new custom function: `missingProperties`, which could get all the variables the template contains.

### CODE REVIEW

File: libraries/botbuilder-dialogs-adaptive/src/generators/multiLanguageGeneratorBase.ts [hunk 1]
```diff
@@ -55,54 +57,121 @@ export abstract class MultiLanguageGeneratorBase<
...
+        }
+
+        return dialogContext.getLocale();
+    }
+
+    private getFallbackLocales(languagePolicy: LanguagePolicy, targetLocale: string): string[] {
+        const fallbackLocales = [];
+        targetLocale = targetLocale.toLowerCase();
+        if (languagePolicy.has(targetLocale)) {
+            languagePolicy.get(targetLocale).forEach((u: string): number => fallbackLocales.push(u));
```
Comment (joshgummersall) [inline libraries/botbuilder-dialogs-adaptive/src/generators/multiLanguageGeneratorBase.ts:157]: `fallbackLocales.push(...laguagePolicy.get(targetLocale))`

File: libraries/botbuilder-dialogs-adaptive/src/generators/multiLanguageGeneratorBase.ts [hunk 2]
```diff
@@ -55,54 +57,121 @@ export abstract class MultiLanguageGeneratorBase<
...
+        if (generators.length === 0) {
+            generators.push(new TemplateEngineLanguageGenerator());
+        }
+
+        const errors: string[] = [];
+        for (const generator of generators) {
+            try {
+                return generator.missingProperties(dialogContext, template, state, options);
+            } catch (e) {
+                errors.push(e);
```
Comment (joshgummersall) [inline libraries/botbuilder-dialogs-adaptive/src/generators/multiLanguageGeneratorBase.ts:101]: Seems like `errors` is unused?

File: libraries/botbuilder-dialogs-adaptive/src/generators/multiLanguageGeneratorBase.ts [hunk 3]
```diff
@@ -6,10 +6,12 @@
  * Licensed under the MIT License.
  */
 
-import { Configurable, Converter, ConverterFactory, DialogContext } from 'botbuilder-dialogs';
+import { Configurable, Converter, ConverterFactory, DialogContext, TurnPath } from 'botbuilder-dialogs';
 import { LanguageGenerator } from '../languageGenerator';
 import { LanguagePolicy, LanguagePolicyConverter } from '../languagePolicy';
 import { languagePolicyKey } from '../languageGeneratorExtensions';
+import { MemoryInterface, Options } from '../../../adaptive-expressions/lib';
```
Comment (joshgummersall) [inline libraries/botbuilder-dialogs-adaptive/src/generators/multiLanguageGeneratorBase.ts:13]: `import ... from 'adaptive-expressions'`

File: libraries/botbuilder-dialogs-adaptive/src/generators/templateEngineLanguageGenerator.ts [hunk 4]
```diff
@@ -9,9 +9,11 @@
 import { Configurable, DialogContext } from 'botbuilder-dialogs';
 import { Resource } from 'botbuilder-dialogs-declarative';
 import { Templates, LGResource, EvaluationOptions } from 'botbuilder-lg';
+import { MemoryInterface, Options } from '../../../adaptive-expressions/lib';
```
Comment (joshgummersall) [inline libraries/botbuilder-dialogs-adaptive/src/generators/templateEngineLanguageGenerator.ts:12]: Import from `adaptive-expressions`

File: libraries/botbuilder-dialogs-adaptive/src/languageGenerator.ts [hunk 5]
```diff
@@ -7,6 +7,7 @@
  */
 
 import { DialogContext } from 'botbuilder-dialogs';
+import { MemoryInterface, Options } from '../../adaptive-expressions/lib';
```
Comment (joshgummersall) [inline libraries/botbuilder-dialogs-adaptive/src/languageGenerator.ts:10]: Import from `adaptive-expressions`

File: libraries/botbuilder-dialogs-adaptive/src/functions/missingPropertiesFunction.ts [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+     */
+    public constructor(context: DialogContext) {
+        super(
+            MissingPropertiesFunction.functionName,
+            MissingPropertiesFunction.function,
+            ReturnType.Array,
+            FunctionUtils.validateUnaryString
+        );
+
+        MissingPropertiesFunction.dialogContext = context;
```
[the remark this answers is not in the document]
Comment (Danieladu) [reply libraries/botbuilder-dialogs-adaptive/src/functions/missingPropertiesFunction.ts:50]: What the function needs is a global dialogContext reference, but it is not available in the function runtime, so dialogContext needs to be passed in when registering. And it also must be static for the registration of the function is static.

File: libraries/botbuilder-dialogs-adaptive/src/generators/multiLanguageGeneratorBase.ts [hunk 7]
```diff
@@ -55,54 +57,116 @@ export abstract class MultiLanguageGeneratorBase<
...
+    private getFallbackLocales(languagePolicy: LanguagePolicy, targetLocale: string): string[] {
+        const fallbackLocales = [];
+        targetLocale = targetLocale.toLowerCase();
+        if (languagePolicy.has(targetLocale)) {
+            fallbackLocales.push(...languagePolicy.get(targetLocale));
         }
 
+        // append empty as fallback to end
+        if (targetLocale !== '' && languagePolicy.has('')) {
+            languagePolicy.get('').forEach((u: string): number => fallbackLocales.push(u));
```
Comment (joshgummersall) [inline libraries/botbuilder-dialogs-adaptive/src/generators/multiLanguageGeneratorBase.ts:157]: Same `push(...)` comment as above.

**Review by joshgummersall — APPROVED** [review]

File: libraries/botbuilder-dialogs-adaptive/etc/botbuilder-dialogs-adaptive.api.md [hunk 8]
```diff
@@ -43,9 +43,11 @@ import { Headers as Headers_2 } from 'node-fetch';
 import { ImportResolverDelegate } from 'botbuilder-lg';
 import { IntExpression } from 'adaptive-expressions';
 import { ListStyle } from 'botbuilder-dialogs';
+import { MemoryInterface } from 'adaptive-expressions';
```
Comment (Danieladu) [inline libraries/botbuilder-dialogs-adaptive/etc/botbuilder-dialogs-adaptive.api.md:46]: @joshgummersall  Thanks josh for this fix!

**Pull request merged**